### ISSUE TITLE
Provide an estimate of remaining (and total) run time

### ISSUE THREAD

**Issue by abensonca** [galacticusorg/galacticus#484]
It would be helpful to provide estimates of remaining (and total) run time. In addition to simply giving some idea of when a job will be finished, this can be useful to estimate the amount of time to request when submitting jobs on clusters. 

The approach could leverage existing meta-analysis that constructs models of how long each tree takes to evolve, and then extrapolate from this if the remaining tree masses are known.

**Comment by abensonca** [comment]
# Issue #484 — Run-time estimation for `evolveForests`: review and plan

*Prepared 2026-07-05. Scope: the `evolveForests` task only. All file references are to current `master`.*

## Summary

Most of the machinery needed for run-time estimation already exists in the codebase, but it is disconnected: per-tree timing is recorded, a cost-model class hierarchy exists, and the evolvers already print a per-tree "estimated time remaining" message when a (non-default) estimator is configured. What is missing is (a) the middle step that turns recorded timing data into a usable cost model, (b) any way for the task level to know how many trees exist and what their masses are, and (c) any task-level (whole-run) progress or ETA reporting. The plan below closes these gaps in phases, with a self-calibrating running estimate that sidesteps the "timing depends on the physics of the model" problem for remaining-time estimates, and an optional prior (from a previous run of the same model) for start-of-run total estimates suitable for sizing cluster job requests.

## 1. What already exists

**Per-tree timing recording** — `mergerTreeOperatorTreeProcessingTimer` (`source/merger_trees/operators/tree_processing_timer.F90`). Hooks the operator callbacks already made by the task (`operatePreConstruction`, `operatePreEvolution`, `operatePostEvolution`; called from `source/tasks/evolve_forests/_class.F90:598,659,869`) and records, per tree: root mass, construction time, evolution time, tree ID, node count, and (optionally) peak memory. Written as raw per-tree arrays to the `metaData/treeTiming` HDF5 group at finalize. No fit is performed; the data are raw. Not enabled by default (the `mergerTreeOperator` default is `null`).

**Cost-model class** — `metaTreeProcessingTime` functionClass (`source/meta/tree_processing_times/_class.F90`) with two methods:

- `time(massTree)` — predicted time to process a tree of given root mass;
- `timeRemaining(tree, timeFinal)` — predicted remaining time for a partially evolved tree.

Implementations:

- `null` (the default) — returns −1 ("no estimate") for both;
- `file` (`source/meta/tree_processing_times/file.F90`) — quadratic polynomial fit of log₁₀(time) vs. log₁₀(mass), with coefficients read from an XML file supplied by the user;
- `totalEvolveTime` (`source/meta/tree_processing_times/total_evolve_time.F90`) — implements `timeRemaining` only: walks the tree's unevolved branch segments, accumulates remaining "work" ∝ t^exponentTime (centrals and satellites separately), and extrapolates elapsed CPU time against work consumed, throttled by an `updateInterval` (default 10 s).

**Per-tree ETA display** — both evolvers already call `timeRemaining` on every tree-walk pass and print "Estimated time remaining to process tree: …s" when the estimate is positive (`source/merger_trees/evolver/standard.F90:579–582`, `source/merger_trees/evolver/threaded.F90:762–767`). With the default `null` estimator this message never appears.

**Work distribution** — `evolveForestsWorkShareClass` with `cyclic`, `stride`, and `firstComeFirstServed` (default) implementations. FCFS hands out forest indices via an atomically incremented MPI-shared counter, giving dynamic load balancing across all MPI ranks and OpenMP threads. Note: the base-class docstring (`source/tasks/evolve_forests/work_share/_class.F90:37`) says balancing is "optionally guided by tree processing time estimates from `metaTreeProcessingTimeClass`", but no implementation actually consults it — that sentence is aspirational/stale.

**Tree census (partially available, not exposed)**:

- *Build path*: `mergerTreeConstructorBuild` computes the full list of tree masses and weights up front (`source/merger_trees/construct/build/_class.F90:347–394`), sorts by mass, and by default processes in descending mass order (`processDescending`, default true) — i.e. most-expensive-first, which combined with FCFS is the current load-balancing strategy.
- *Read path*: `mergerTreeConstructorRead` opens files lazily; per open file the importer reports forest count and per-forest node counts and weights before import, but root masses only after import, and totals across files only once every file has been opened.
- In both cases this information is **private to the constructor**: the `mergerTreeConstructorClass` interface exposes only `construct(treeNumber, finished)` — a pull-based stream. Neither the task loop nor the work share ever learns the total tree count or mass list.

**Other relevant pieces**:

- One existing consumer of the `file` cost model: `mergerTreeBuildMassDistributionStllrMssFnctn` uses `time(mass)` to compute a compute-time-optimal tree-mass sampling density (`source/merger_trees/construct/build/masses/distribution/stellar_mass_function.F90:422–424`).
- `displayCounter()` (`source/display/_module.F90:360–417`) provides a percentage-complete progress display, used by tabulation tasks but not by tree evolution.
- `walltimeMaximum` parameter on the task: stops claiming new forests when the wall-time budget is exhausted — a natural companion to an ETA (warn if projected time exceeds the budget).
- The threaded evolver already stores each node's last ODE-evolve wall time in a meta-property and dispatches nodes most-expensive-first — within-tree cost balancing exists.

## 2. The gaps

1. **No calibration tooling.** Nothing in the repo (Fortran or scripts) fits the recorded `metaData/treeTiming` data to produce the coefficients XML that `metaTreeProcessingTimeFile` consumes. The recording side and the consumption side exist; the middle step is entirely manual and undocumented.
2. **No census API.** Total tree count and masses are known inside the build constructor but invisible to the task, so no "N of M trees complete" or sum-over-remaining-trees estimate is possible today.
3. **No task-level reporting.** All existing timing displays are per-tree. There is no whole-run total estimate, no remaining-time estimate, no percentage complete, and no end-of-run summary that would help size the next job submission.
4. **Single-large-tree estimation is weak.** `totalEvolveTime` exists but uses `CPU_Time()/threadCount` (crude under the threaded evolver), a heuristic work model (t^0.5), and is off by default so nobody sees it.

## 3. Key challenges (and how the plan addresses them)

**Model dependence.** As noted in the issue, the time to evolve a tree depends on the physics of the specific model (as well as mass, resolution, output count, …). A cost model fitted to one model is wrong for another. Two complementary answers:

- *For remaining-time estimates*: self-calibrate from the run itself. The trees completed so far in *this* run, with *this* physics, are the ideal calibration sample — no prior needed. With descending-mass processing the expensive trees complete first, so the estimate converges quickly and the extrapolation to remaining (cheaper) trees is benign.
- *For start-of-run total estimates* (the cluster-job-sizing use case): a prior from a previous run of the same (or similar) model is unavoidable — there is no information at t=0 otherwise. Make producing and consuming that prior nearly frictionless (Phase 1), and clearly label the estimate as conditional on the calibration model.

**Scatter at fixed mass.** Per-tree times at fixed root mass scatter substantially (branching history, stochastic physics). The fit should record the residual scatter, and estimates should be quoted with an uncertainty (e.g. "estimated remaining: 3.2 h ± 0.4 h"). Summing over many trees averages the scatter down; the single-tree case is where it is irreducible.

**The single-large-tree case.** A mass-based model predicts total time but says nothing about progress within the tree. Here the `totalEvolveTime` approach (work-integral extrapolation) is the right tool and just needs hardening and enabling. The task's per-output-time loop (`source/tasks/evolve_forests/_class.F90:688–844`) also gives a coarse, honest progress marker for free ("output 3 of 10 written"), though work per output interval is highly non-uniform, so it should inform, not anchor, the estimate.

**Parallelism.** Total wall time ≈ total work / effective workers, degraded by end-of-run load imbalance. Within one process, OpenMP-shared accumulators are cheap. Across MPI ranks, the FCFS counter already provides a globally consistent "trees claimed" count essentially for free; aggregating actual times across ranks is more work and can be deferred (per-rank ETAs under FCFS balance are nearly identical anyway).

**Read-path trees.** Masses are not known up front, but per-forest node counts are (per open file), and node count is arguably a better cost predictor than mass anyway (the timer already records both). The cost model should eventually support node count as an alternative predictor; initially the read path can degrade to self-calibrated per-tree averages.

## 4. How to present the information (UX)

Proposed reporting surface, gated on verbosity as usual:

1. **Start of run** (only when a prior cost model and a census are available): one message with estimated total work (core-hours), estimated wall time given the worker count, and the calibration source. If `walltimeMaximum` is set and the projection exceeds it, warn.
2. **During the run, many-tree case**: a throttled message (new parameter, e.g. `timeIntervalReportProgress`, default ~60 s; wall-clock-throttled, not per-tree, so it works for both many small trees and few large ones): trees completed / total (if known), percentage of estimated work complete, elapsed, estimated remaining ± uncertainty. Optionally the `displayCounter` percentage bar for interactive runs.
3. **During the run, single-large-tree case**: the existing evolver message ("Estimated time remaining to process tree") becomes visible once `totalEvolveTime` (or successor) is usable by default; the task loop additionally reports at each output-time boundary.
4. **End of run**: a summary — total wall time, total core-time, tree count, mean/max tree time — phrased to be directly useful for the next job submission ("this model: X core-hours for N trees").

Messages, not files, are the primary surface; but the fitted cost model should also land in the output HDF5 (see Phase 1) so the *next* run can consume it.

## 5. Implementation plan

**Phase 1 — close the calibration loop.** Make `treeProcessingTimer` perform the quadratic log-log fit itself at finalize (it already holds all the data) and write the coefficients (plus residual scatter and mass range) into `metaData/treeTiming`, alongside the raw data. Extend `metaTreeProcessingTimeFile` to accept either the existing XML format or a Galacticus HDF5 output file, so `<metaTreeProcessingTime value="file"><fileName value="previousRun.hdf5"/></...>` "just works". Optionally add a small Python helper (`scripts/analysis/`) for fitting/combining timing data across multiple outputs. Self-contained; no behavioral change to evolution.

**Phase 2 — census API.** Add optional self-describing methods to `mergerTreeConstructorClass`, e.g. `treeCount()` (−1 if unknown) and `treeMasses()` (empty if unknown), with defaults meaning "unknown" so every implementation degrades gracefully. Implement fully in `build` (trivial — the arrays already exist); implement incrementally in `read` (counts/node counts per opened file). Purely additive.

**Phase 3 — task-level progress and ETA.** In `evolveForestsPerform`: shared accumulators (OpenMP-safe) of trees completed and (predicted, actual) time pairs; at start, the Phase-1/2-powered total estimate; during the run, a throttled progress/ETA message using the self-calibration scheme — with a prior, rescale predictions by the running median of actual/predicted; without one, fit on the fly once enough trees have completed; remaining = corrected Σ time(Mᵢ) over unprocessed trees / worker count. End-of-run summary. `walltimeMaximum` overrun warning. MPI: report from each rank (or master only) initially; cross-rank aggregation deferred to Phase 5.

**Phase 4 — single-large-tree hardening.** Fix `totalEvolveTime` to use wall clock consistently (it currently mixes `CPU_Time()/threadCount`), validate/tune the work exponent against recorded timing data, add a fraction-of-work-complete figure to its output, and consider making it (or a successor) the default `metaTreeProcessingTime` — low risk, since its only effect is the display message. Add per-output-boundary progress in the task loop.

**Phase 5 — extensions (optional, separable).** (a) Cost-model-guided work sharing, making the stale docstring in `work_share/_class.F90:37` true — e.g. a work share that orders/assigns forests by predicted cost; mostly redundant with descending-mass + FCFS, so justify with measurements first. (b) Node-count-based cost predictor for the read path. (c) Cross-MPI-rank ETA aggregation. (d) A "dry-run estimate" mode (estimate and exit) for job sizing without starting evolution.

## 6. Open questions

1. Should `treeProcessingTimer` be enabled by default (its overhead is a clock read and a node-count walk per tree), so that *every* run produces calibration data and the "previous run of the same model" prior is always available?
2. Predictor variable: root mass (natural for the build path), node count (natural for the read path, and available pre-evolution), or both? Suggest fitting both in Phase 1 since the timer records both.
3. Is per-process (OpenMP-only) ETA acceptable for the first release, deferring MPI aggregation?
4. Default cadence and verbosity level for the progress messages; should the interactive `displayCounter` bar be used at all?
